### ISSUE TITLE
Allow dynamic reconfiguration of webhooks

### ISSUE THREAD

**Issue by mattmoor** [knative/pkg#782]
tl;dr I'd like to be able to use our webhook infrastructure to process dynamically discovered implementations of a give duck type.

[Here's an example](https://github.com/mattmoor/boo-maps/blob/5ed68ca4e0f45e01d45ccebaddefecea263b6f9d/cmd/webhook/main.go#L122-L131) that uses a duck type with our webhook (not dynamic), but the last entry clearly hints at why being able to dynamically reconfigure the webhook would be interesting (though it is unclear what it should key off of for discovery).

[Here's another example](https://github.com/mattmoor/foo-binding#extensibility), where the method of dynamic discovery is clearer.  Essentially, as new types are referenced from binding resources, the webhook would adapt to process them.

**Comment by mattmoor** [comment]
This is going to sounds somewhat like a hammer seeing everything as nails, but my initial take on this is that we should push the currently very ad hoc webhook registration process towards the reconciler pattern.  This would naturally give us a pattern that lets the webhook respond and adapt to changes.

Related to https://github.com/knative/pkg/issues/450 (and it's deeper sentiment) we should refactor the webhook to tease apart a few logical parts to break up the monolithic webhook package to try to fit this pattern.

**Comment by mattmoor** [comment]
Note that this gets more interesting (and the logical parts clearer) as the number and kind of webhooks expand.  We now have Mutating for types (defaulting, validation), Validating for configmaps (validation), and should soon have Conversion for types.

**Comment by vagababov** [comment]
How do we do validation and the defaulting the currently a rely on explicitly calling the methods on the types.

**Comment by nimakaviani** [comment]
Defaulting can probably be encoded in the same resource that holds the reference bindings, eg by pointing to a config map or providing key/values?

But to @vagababov 's point, what about validation on fields that fall outside what is validated with the duck type?

**Comment by n3wscott** [comment]
We could bring in the OpenAPI of the duck and use that as a first pass.
